src/load.jl

```diff
@@ -0,0 +1,217 @@
+#     ϵ = get(attrs, :epsilon, 1f-5)
+#     momentum = get(attrs, :momentum, 9f-1)
+#     training_mode = Bool(get(attrs, :training_mode, 0))
+#     res = push_call!(tape, batch_norm, args..., ϵ, momentum, training_mode)
+#     if training_mode
+#         y = push_call!(tape, getfield, 1)
+#         μ_new = push_call!(tape, getfield, 2)
+#         σ²_new = push_call!(tape, getfield, 3)
+#         return y, μ_new, σ²_new
+#     else
+#         return res
+#     end
+# end
+
+
+# function load_node!(tape::Tape, ::OpConfig{:ONNX, :GlobalAveragePool}, args::VarVec, attrs::AttrDict)
+#     return push_call!(tape, global_average_pool, args...)
+# end
+
+
+###############################################################################
+#                                    API                                      #
+###############################################################################
+
+
+"""
+    load(io::IO, model_args...; backends=[:ONNX], exec::Bool=true)
+    load(filename::String, model_args...; backends=[:ONNX], exec::Bool=true)
+
+Load an ONNX model as a Ghost.Tape. The way a particular ONNX node is deserialized is
+controlled by methods of [load_node!](@ref) dispatched by backend and node's op_type.
+
+`backends` parameter can be used to customize the loading process.
+
+`exec` parameter instructs the loader to execute every added operation just after
+the addition, making the debugging easier. Default is `true`.
+
+See also: [save!](@ref)
+"""
+function load(io::IO, model_args...; backends=[:ONNX], exec::Bool=true)
+    onnx_model = readproto(io, ModelProto());
+    g = onnx_model.graph;
+    tape = Tape(ONNXCtx(backends; exec=exec))
+    # create map of initializers
+    init_vals = Dict{String, Any}()
+    for init in g.initializer
+        # TODO: consider non-array inputs
+        init_vals[init.name] = array(init)
+    end
+    # load inputs
+    arg_idx = 1
+    for inp in g.input
+        val = get(init_vals, inp.name, missing)
+        if val === missing && exec == true
+            val = model_args[arg_idx]
+            arg_idx += 1
+        end
+        v = push!(tape, Input(val))
+        tape.c.name2var[inp.name] = v
+    end
+    # load nodes
+    op_configs = [meth.sig.parameters[3] for meth in methods(load_node!).ms]
+    op_configs = [config for config in op_configs if config <: OpConfig]
+    for nd in g.node
+        success = false
+        for backend in tape.c.backends
+            # check if there's an implementation for this op in this backend
+            if OpConfig{backend, Symbol(nd.op_type)} in op_configs
+                load_node!(tape, nd, backend)
+                success = true
+                break
+                @debug "Loaded $(nd.op_type) using backend $(backend)"
+            end
+        end
+        success || error("Couldn't load node for $(nd.op_type), " *
+                        "tried the following backends: $(tape.c.backends)")
+    end
```

Comment:
In line with @DrChainsaw's comments, is there a need to store all the possible values defined for `OpConfig` + `load_node!`, then dispatch on that config? Why not define
```julia
load_node!(::Tape, ::OpConfig{BE, Op}, ::VarVec, ::AttrDict) = missing
```
Then we directly try to dispatch like:
```julia
for nd in g.node
  success = false
  for backend in tape.c.backends
    if !ismissing(load_node!(tape, nd, backend))
      success = true
      @debug "Loaded $(nd.op_type) using backend $(backend)"
      break
    end
  end
  success || error("Couldn't load node for $(nd.op_type), " *
                   "tried the following backends: $(tape.c.backends)")
end
```
The dispatch for `load_node!(tape, nd, backend)` would need to skip the `missing` returns and pass them through.

Reply:
Yeah, originally I used similar logic, then changed the API of `load_node!()` a couple of times that made it not applicable, came up with 2 weird solutions (this is the second), but now we can indeed return to this solution. Thanks for the code suggestion!